pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,288 @@
+            signature of the Bloq as shown in the example.
+
+    Raises:
+        TypeError: bloq must be an instance of :code:`~.Bloq`.
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    >>> from qualtran.bloqs.basic_gates import CNOT
+
+    >>> qualtran_cnot = qml.FromBloq(CNOT(), wires=[0, 1])
+    >>> qualtran_cnot.matrix()
+    array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+       [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j]])
+
+    A simple example showcasing how to use ``qml.FromBloq`` inside a device:
+
+    .. code-block::
+
+        from qualtran.bloqs.phase_estimation import RectangularWindowState, TextbookQPE
+        from qualtran.bloqs.chemistry.trotter.ising import IsingXUnitary, IsingZZUnitary
+        from qualtran.bloqs.chemistry.trotter.trotterized_unitary import TrotterizedUnitary
+
+        # Parameters for the TrotterizedUnitary
+        nsites = 5
+        j_zz = 2
+        gamma_x = 0.1
+        dt = 0.01
+        indices = (0, 1, 0)
+        coeffs = (0.5 * gamma_x, j_zz, 0.5 * gamma_x)
+        zz_bloq = IsingZZUnitary(nsites=nsites, angle=2 * dt * j_zz)
+        x_bloq = IsingXUnitary(nsites=nsites, angle=0.5 * 2 * dt * gamma_x)
+        trott_unitary = TrotterizedUnitary(
+            bloqs=(x_bloq, zz_bloq), indices=indices, coeffs=coeffs, timestep=dt
+        )
+
+        # Instantiate the TextbookQPE and pass in our unitary
+        textbook_qpe = TextbookQPE(trott_unitary, RectangularWindowState(3))
+
+        # Execute on device
+        dev = qml.device("default.qubit")
+        @qml.qnode(dev)
+        def circuit():
+            qml.FromBloq(textbook_qpe, wires=range(textbook_qpe.signature.n_qubits()))
+            return qml.probs(wires=[5, 6, 7])
+
+    >>> circuit()
+    array([0.94855734, 0.01291602, 0.00431044, 0.00267219, 0.00237645,
+    0.00290337, 0.00526717, 0.02099701])
+
+    .. details::
+        :title: Usage Details
+
+        The decomposition of a bloq wrapped in ``qml.FromBloq`` may use more wires than expected.
+        For example, when we wrap Qualtran's ``CZPowGate``, we get
+
+        >>> from qualtran.bloqs.basic_gates import CZPowGate
+        >>> qml.FromBloq(CZPowGate(0.468, eps=1e-11), wires=[0, 1]).decomposition()
+        [FromBloq(And, wires=Wires([0, 1, 'alloc_free2'])),
+        FromBloq(Z**0.468, wires=Wires(['alloc_free2'])),
+        FromBloq(And†, wires=Wires([0, 1, 'alloc_free2']))]
+
+        This is not a bug. It is due to the fact that the decomposition of ``CZPowGate`` as
+        defined in qualtran allocates and frees a wire all in the same bloq. In this situation,
+        PennyLane automatically allocates this wire under the hood, and that additional wire is
+        named ``alloc_free{idx}``, where idx is the total length of wires. For now, due to technical
+        limitations these wires cannot be accessed manually, or mapped to a different name.
```

Comment:
```suggestion
        This sharp bit is expected behaviour because the decomposition of ``CZPowGate`` as
        defined in qualtran allocates and frees a wire all in the same ``bloq``. In this situation,
        PennyLane automatically allocates this wire under the hood, and that additional wire is
        named ``alloc_free_{idx}``, where ``idx`` is used for bookkeeping index of all such
        ghost wires. Due to the current limitations of PennyLane's wire management in dealing
        with such wires, these cannot be accessed manually or mapped to a different label.
```